### PR TITLE
Constant liar for `TPESampler`

### PR DESCRIPTION
**Pull request by hvy**
## Motivation

Fixes https://github.com/optuna/optuna/issues/892.

See also https://github.com/optuna/optuna/pull/2217.

## Description of the changes

Introduces a `constant_liar: bool` parameter to `TPESampler` to allow turning constant liar on.

When this argument is `False` (default), the previous behavior is kept. If `True`, the values of the running states are penalized with `float("inf")` or `-float("inf")` depending on the direction. A more flexible `Union[bool, Callable[[Sequence[float]], float]]` would allow arbitrary values for penalization but I left it out to keep the behavior simple avoiding confusion with how the direction is impacting the final value.

### Benchmarks

https://colab.research.google.com/drive/1_JaQGl_sCbbujTdZbd9-_MQ9Bj8dROcF?usp=sharing

### Notes

_Note that this PR may conflict with the ongoing TPE refactoring (https://github.com/optuna/optuna/pull/2618)._

### CODE REVIEW

**Comment by HideakiImamura** [comment]
@keisuke-umezawa Could you review this PR if you have time?

File: tests/samplers_tests/tpe_tests/test_sampler.py
Comment (hvy) [inline tests/samplers_tests/tpe_tests/test_sampler.py:977]: Open for ideas if there's a better way to test. This test uses private APIs and doesn't test the propagation of the `constant_liar` argument from `__init__` to observation pair extraction.

File: optuna/samplers/_tpe/sampler.py
Comment (HideakiImamura) [inline optuna/samplers/_tpe/sampler.py:944]: The size of the constant is fixed to infinity, but it could be made adjustable (not a strong opinion).

File: optuna/samplers/_tpe/sampler.py
Comment (HideakiImamura) [inline optuna/samplers/_tpe/sampler.py:960]: This function is merged in #2618. I think it's good to wait for #2618 to merge this PR.

File: tests/samplers_tests/tpe_tests/test_sampler.py
Comment (HideakiImamura) [reply tests/samplers_tests/tpe_tests/test_sampler.py:977]: This might be out of the scope of this PR, but I have an idea not for the test but for verification of the algorithm itself. How about plotting the density of GMM like the following setting?

- Consider low-dimensional objective function such as the rosenblock function. 
- For `x: float`, the value of the density of GMM is calculated by `TPESampler._gmm_log_pdf(samples=[x], ...)`.

**Review by HideakiImamura — COMMENTED** [review]
Thanks for the PR! I'm very excited to see this algorithm in the Optuna. I have several comments. PTAL.

File: optuna/samplers/_tpe/sampler.py
Comment (hvy) [reply optuna/samplers/_tpe/sampler.py:944]: I agree. I tried to address that in the description.

File: tests/samplers_tests/tpe_tests/test_sampler.py
Comment (hvy) [reply tests/samplers_tests/tpe_tests/test_sampler.py:977]: Right, we could perhaps do that to check the algorithm. But I'm wondering if that isn't already verifiable using the notebook in the description. Do you think we can write unit tests based on this?

File: optuna/samplers/_tpe/sampler.py
Comment (keisuke-umezawa) [inline optuna/samplers/_tpe/sampler.py:217]: In nii, they can configure constant_liar by following arguments:

> classArgs requirements:

> optimize_mode (maximize or minimize, optional, default = maximize) - If 'maximize', tuners will try to maximize metrics. If 'minimize', tuner will try to minimize metrics.
>parallel_optimize (bool, optional, default = False) - If True, TPE will use the Constant Liar algorithm to optimize parallel hyperparameter tuning. Otherwise, TPE will not discriminate between sequential or parallel situations.
>constant_liar_type (min or max or mean, optional, default = min) - The type of constant liar to use, will logically be determined on the basis of the values taken by y at X. There are three possible values, min{Y}, max{Y}, and mean{Y}.

https://github.com/microsoft/nni/blob/e85f029bd4e4b1bdf3e679893fb6447e4d6b2c79/docs/en_US/Tuner/HyperoptTuner.rst

Comment (hvy) [reply optuna/samplers/_tpe/sampler.py:217]: Do mean mean we should do that in Optuna as well? I tried to address that in the PR description.

File: tests/samplers_tests/tpe_tests/test_sampler.py
Comment (HideakiImamura) [reply tests/samplers_tests/tpe_tests/test_sampler.py:977]: No, I don't think we need to go that far with unit tests. Indeed, as you said, we can see that various points are being evaluated at once, as shown in the colab above.

I thought this would be a good opportunity as the request to plot KDE values has been raised by users before, but this seems to be beyond the scope of this PR. (FYI: https://github.com/optuna/optuna/issues/2600#issuecomment-824663938)

File: optuna/samplers/_tpe/sampler.py
Comment (HideakiImamura) [reply optuna/samplers/_tpe/sampler.py:217]: Looks reasonable. @hvy How about adding `constant_liar_type` argument?

Comment (hvy) [reply optuna/samplers/_tpe/sampler.py:217]: Do you mean introducing another parameter in addition to `constant_liar` ? One should be enough since it includes the boolean information. However, again, I'm worried that it'll be more confusing than useful for many users.

Comment (HideakiImamura) [reply optuna/samplers/_tpe/sampler.py:217]: > One should be enough since it includes the boolean information.

You are right. It's enough to include the type of constant liar information in `constant_liar`. 

> I'm worried that it'll be more confusing than useful for many users.

I see. It's OK for me to restrict the type of constant liar to `inf` in this PR. I think it will be enough if we work on it after the users request to change the type of constant liar.

**Review by HideakiImamura — APPROVED** [review]
LGTM! (All my concerns are resolved. Thanks, @hvy!)

**Comment by HideakiImamura** [comment]
Note: we will resolve conflicts in #2618 after merging this PR.

File: optuna/samplers/_tpe/sampler.py
Comment (keisuke-umezawa) [reply optuna/samplers/_tpe/sampler.py:217]: Sorry, it is just FYI.

And, especially, I noticed that they named the argument from the functionality `parallel_optimize`, not from the implementation `constant_liar`.
 
The argument `constant_liar` is a little bit difficult for users to recognize that the functionality is for distributed optimization, although they can easily understand that if they read docstrings.

In my opinion, we can use `constant_liar` here, because users will read the docstrings thoroughly if they want to use the functionality.

**Review by keisuke-umezawa — APPROVED** [review]
LGTM! Thank you for the implementation!

**Pull request merged**
